lib/pathmatch.cpp

```diff
@@ -19,74 +19,91 @@
-    else
-        finddir = findpath;
-
-    const bool is_absolute = Path::isAbsolute(path);
-
-    // TODO: align the match logic with ImportProject::ignorePaths()
-    for (auto i = mPaths.cbegin(); i != mPaths.cend(); ++i) {
-        const std::string pathToMatch((!is_absolute && Path::isAbsolute(*i)) ? Path::getRelativePath(*i, mWorkingDirectory) : *i);
-
-        // Filtering directory name
-        if (endsWith(pathToMatch,'/')) {
-            if (pathToMatch.length() > finddir.length())
-                continue;
-            // Match relative paths starting with mask
-            // -isrc matches src/foo.cpp
-            if (finddir.compare(0, pathToMatch.size(), pathToMatch) == 0)
-                return true;
-            // Match only full directory name in middle or end of the path
-            // -isrc matches myproject/src/ but does not match
-            // myproject/srcfiles/ or myproject/mysrc/
-            if (finddir.find("/" + pathToMatch) != std::string::npos)
-                return true;
-        }
-        // Filtering filename
-        else {
-            if (pathToMatch.length() > findpath.length())
-                continue;
-            // Check if path ends with mask
-            // -ifoo.cpp matches (./)foo.c, src/foo.cpp and proj/src/foo.cpp
-            // -isrc/file.cpp matches src/foo.cpp and proj/src/foo.cpp
-            if (findpath.compare(findpath.size() - pathToMatch.size(), findpath.size(), pathToMatch) == 0)
-                return true;
+    std::string regex_string;
+
+    for (auto p : paths) {
+        if (p.empty())
+            continue;
+
+        if (!regex_string.empty())
+            regex_string.push_back('|');
+
+        p = Path::fromNativeSeparators(p);
+
+        if (p.front() == '.')
+            p = Path::getCurrentPath() + "/" + p;
+
+        if (Path::isAbsolute(p)) {
+            p = Path::simplifyPath(p);
+
+            if (p.back() == '/')
+                regex_string.append("^" + translate(p));
+            else
+                regex_string.append("^" + translate(p) + "$");
+        } else {
+            if (p.back() == '/')
+                regex_string.append("/" + translate(p));
+            else
+                regex_string.append("/" + translate(p) + "$");
         }
     }
-    return false;
+
+    if (regex_string.empty())
+        regex_string = "^$";
+
+    if (caseSensitive)
+        mRegex = std::regex(regex_string, std::regex_constants::extended);
+    else
+        mRegex = std::regex(regex_string, std::regex_constants::extended | std::regex_constants::icase);
 }
 
-std::string PathMatch::removeFilename(const std::string &path)
+bool PathMatch::match(const std::string &path) const
 {
-    const std::size_t ind = path.find_last_of('/');
-    return path.substr(0, ind + 1);
+    std::string p;
+    std::smatch m;
+
+    if (Path::isAbsolute(path))
+        p = Path::simplifyPath(path);
+    else
+        p = Path::simplifyPath(Path::getCurrentPath() + "/" + path);
```

Comment:
I have the feeling it would be good to create a member `mCurrentPath` that is used instead of the function call. It will be good for unit testing purposes. And avoids repeated `getcwd` system calls.